### DISCUSSION TITLE
Cluster for JSX elements?

### DISCUSSION THREAD

**Discussion by Cragsmann** [uber/react-map-gl#2350] [Q&A]
Hello, I am wondering if there is a way to cluster custom jsx markers. Until now I have found only a html clusters with source and layers or supercluster. Thank you for your answers.

**Comment by khmm12** [comment]
Greetings :wave:

You can cluster custom JSX markers – the trick is to run supercluster against the current viewport and render the result as regular `<Marker>` components with any JSX inside, instead of going through sources/layers.

I maintain a small library that wires this up for react-map-gl: [react-map-gl-supercluster](https://github.com/khmm12/react-map-gl-supercluster). It exposes a `useSupercluster` hook that recomputes clusters on the current bounds/zoom, and you render them however you like:

```tsx
import Map, { Marker } from 'react-map-gl/mapbox'
import { isCluster, useSupercluster } from 'react-map-gl-supercluster/mapbox'

function Markers({ points }) {
  const { clusters } = useSupercluster(points)

  return clusters.map((cluster) => {
    const [longitude, latitude] = cluster.geometry.coordinates

    return isCluster(cluster) ? (
      <MyClusterPin key={`cluster-${cluster.properties.cluster_id}`} longitude={longitude} latitude={latitude} />
    ) : (
      <MyPin key={cluster.properties.item.id} longitude={longitude} latitude={latitude} />
    )
  })
}
```

Entrypoints mirror react-map-gl v8: `/mapbox`, `/maplibre` and `/mapbox-legacy`. Live demo: [StackBlitz](https://stackblitz.com/github/khmm12/react-map-gl-supercluster)
